language-plutus-core/src/Language/PlutusCore/Normalize.hs

```diff
@@ -59,25 +70,34 @@ added to environments and normalization instantiates all variables presented in
 -- | Normalize a 'Type' in the 'NormalizeTypeM' monad.
 normalizeTypeM
     :: HasUnique (tyname ()) TypeUnique
-    => Type tyname () -> NormalizeTypeM tyname (NormalizedType tyname ())
-normalizeTypeM (TyForall ann name kind body) = TyForall ann name kind <<$>> normalizeTypeM body
-normalizeTypeM (TyFix ann name pat)          = TyFix ann name <<$>> normalizeTypeM pat
-normalizeTypeM (TyFun ann dom cod)           =
+    => Type tyname () -> NormalizeTypeM tyname a (NormalizedType tyname ())
+normalizeTypeM (TyForall ann name kind body) = do
+    normalizeTypeStep
```

Comment:
My reading of https://iohk.myjetbrains.com/youtrack/issue/CGP-276 is that we're interested in the number of reduction steps that we do. These arent't reduction steps, they're morally the identity.

To put it another way, normalizing a type that is already normal should not use any steps, which currently it will.

Reply:
In practice what I think this means is that we want to bound the number of *substitutions* that we do.

Reply:
Yes, it is important that normalising a normal type doesn't use any steps.